README.md

```diff
@@ -24,21 +23,129 @@ Join us to help define the direction and implementation of this project!
+
+   * cd to the repository directory
+   * run `make uninstall`
+
+## How to Use Gatekeeper
+
+### Constraint Templates
+
+Before you can define a constraint, you must first define a `ConstraintTemplate`, which describes both the Rego that enforces the constraint and the schema of the constraint. The schema of the constraint allows an admin to fine-tune the behavior of a constraint, much like arguments to a function.
+
+Here is an example constraint template that requires all labels described by the constraint to be present:
+
+```yaml
+apiVersion: templates.gatekeeper.sh/v1alpha1
+kind: ConstraintTemplate
+metadata:
+  name: k8srequiredlabels
+spec:
+  crd:
+    spec:
+      names:
+        kind: K8sRequiredLabels
+        listKind: K8sRequiredLabelsList
+        plural: k8srequiredlabels
+        singular: k8srequiredlabels
+      validation:
+        # Schema for the `parameters` field
+        openAPIV3Schema:
+          properties:
+            labels:
+              type: array
+              items: string
+  targets:
+    - target: admission.k8s.gatekeeper.sh
+      rego: |
+        package k8srequiredlabels
+
+        deny[{"msg": msg, "details": {"missing_labels": missing}}] {
+          provided := {label | input.review.object.metadata.labels[label]}
+          required := {label | label := input.constraint.spec.parameters.labels[_]}
+          missing := required - provided
+          count(missing) > 0
+          msg := sprintf("you must provide labels: %v", [missing])
+        }
+```
+
+### Constraints
+
+Constraints are then used to inform Gatekeeper that the admin wants a ConstraintTemplate to be enforced, and how. This constraint uses the `K8sRequiredLabels` constraint above to make sure the `gatekeeper` label is defined on all namespaces:
+
+```yaml
+apiVersion: constraints.gatekeeper.sh/v1alpha1
+kind: K8sRequiredLabels
+metadata:
+  name: ns-must-have-gk
+spec:
+  match:
+    kinds:
+      - apiGroups: [""]
+        kinds: ["Namespace"]
+  parameters:
+    labels: ["gatekeeper"]
+```
+
+Note the `match` field, which defines the scope of objects to which a given constraint will be applied. It supports the following matchers:
+
+   * `kinds` accepts a list of objects with `apiGroups` and `kinds` fields that list the groups/kinds of objects to which the constraint will apply. If multiple groups/kinds objects are specified, only one match is needed for the resource to be in scope.
+   * `namespaces` is a list of namespace names. If defined, a constraint will only apply to resources in a listed namespace.
+   * `labelSelector` is a standard Kubernetes label selector.
+
+Note that if multiple matchers are specified, a resource must satisfy each top-level matcher (`kinds`, `namespaces`, etc.) to be in scope. Each top-level matcher has its own semantics for what qualifies as a match. An empty matcher is deemed to be inclusive (matches everything).
+
+### Replicating Data
+
```

Comment:
Perhaps explain why Kubernetes data needs to be replicated into OPA first before going into the how.
```suggestion
Replicating Kubernetes data into OPA is required to ensure the audit feature works properly and to support policies that require pair-wise search. For example: A policy that enforces the uniqueness of an ingress's hostname among all ingresses in the cluster.
```

Reply:
Good point. Adding a bit more info around replication generally.